File: face_comparsion_cpp/src/main/java/com/netposa/poseidon/face/util/LoadPropers.java
package com.netposa.poseidon.face.util;

import org.apache.log4j.Logger;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

/**
 * Created by Nukil on 2017/4/
 */
public class LoadPropers {
    private Properties properties;
    private Properties kafkaProperties;
    private Logger logger = Logger.getLogger(LoadPropers.class);

    private LoadPropers() {
    }

    private static LoadPropers instance = null;

    public static LoadPropers getSingleInstance() {
        if (null == instance) {
            synchronized (LoadPropers.class) {
                if (null == instance) {
                    instance = new LoadPropers();
                }
            }
        }
        return instance;
    }

    private Properties loadProperties(String filename) {
        ClassLoader classLoader = this.getClass().getClassLoader();
        Properties tmpProperties = new Properties();
        InputStream configStream = classLoader.getResourceAsStream(filename + ".properties");
        try {
            tmpProperties.load(configStream);
        } catch (IOException e) {
            logger.error(e.getMessage());
        } finally {
            try {
                configStream.close();
            } catch (IOException x) {
                logger.error(x.getMessage());
            }
        }
        return tmpProperties;
    }

    public Properties getProperties(String filename) {
        Properties rProperties;
        switch (filename) {
            case "server":
                if (null == properties) {
                    synchronized (LoadPropers.class) {
                        if (null == properties) {
                            properties = loadProperties(filename);
                        }
                    }
                }
                rProperties = properties;
                break;
            case "kafka-consumer":
                if (null == kafkaProperties) {
                    synchronized (LoadPropers.class) {
                        if (null == kafkaProperties) {
                            kafkaProperties = loadProperties(filename);
                        }
                    }
                }
                rProperties = kafkaProperties;
                break;
            default:
                rProperties = loadProperties(filename);
        }
        return rProperties;
    }
}

File: face_comparsion/src/main/java/com/netposa/poseidon/face/init/LoadPropers.scala
package com.netposa.poseidon.face.init

import java.util.Properties

import scala.collection.JavaConversions.asScalaSet
import org.apache.commons.lang.StringUtils
import org.slf4j.{Logger, LoggerFactory}
import com.netposa.poseidon.face.init.LoadPropers.LOG


/**
  * 加载配置文件为java.util.Properties对象
  * 将该类修改为单例模式
  */
class LoadPropers private() {
    def load(): Properties = {
        val class_loader = this.getClass.getClassLoader
        LOG.info("load server.properties config file by class path")
        val tmp_props = new Properties()
        val server_conf_stream = class_loader.getResourceAsStream("face-server.properties")
        try {
            tmp_props.load(server_conf_stream)
            val tmp = tmp_props.keySet().map(key => {
                val value = StringUtils.trimToNull(tmp_props.getProperty(key.toString))
                LOG.info("config server properties key is %s value is %s".format(key.toString, value))
                (key.toString, value)
            }).toMap

            val server_props = new Properties()
            tmp.foreach(row => {
                if (row._1 != null && row._2 != null) {
                    server_props.put(row._1, row._2)
                }
            })
            server_props
        } finally {
            try {
                server_conf_stream.close()
            } catch {
                case _: Throwable =>
            }
        }
    }
    def load(fileName: String): Properties = {
        val class_loader = this.getClass.getClassLoader
        LOG.info("load server.properties config file by class path")
        val tmp_props = new Properties()
        val server_conf_stream = class_loader.getResourceAsStream(fileName)
        try {
            tmp_props.load(server_conf_stream)
            val tmp = tmp_props.keySet().map(key => {
                val value = StringUtils.trimToNull(tmp_props.getProperty(key.toString))
                LOG.info("config server properties key is %s value is %s".format(key.toString, value))
                (key.toString, value)
            }).toMap

            val server_props = new Properties()
            tmp.foreach(row => {
                if (row._1 != null && row._2 != null) {
                    server_props.put(row._1, row._2)
                }
            })
            server_props
        } finally {
            try {
                server_conf_stream.close()
            } catch {
                case _: Throwable =>
            }
        }
    }
}

//修改成单例
object LoadPropers {
    val LOG: Logger = LoggerFactory.getLogger(LoadPropers.getClass)
    var properties: Properties = _

    def getProperties(): Properties = {
        if (properties == null) {
            properties = new LoadPropers().load()
        }
        properties
    }
    def getProperties(fileName: String): Properties = {
        new LoadPropers().load(fileName + ".properties")
    }
}